File: app/src/main/java/com/ngengs/android/popularmovies/apps/utils/networks/NetworkHelpers.java
package com.ngengs.android.popularmovies.apps.utils.networks;

import android.app.Application;

import com.ngengs.android.popularmovies.apps.globals.Values;

import java.util.concurrent.TimeUnit;

import okhttp3.Cache;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by ngengs on 7/4/2017.
 */

@SuppressWarnings({"DefaultFileTemplate", "WeakerAccess"})
public class NetworkHelpers {
    private static final long CACHE_SIZE = 10 * 1024 * 1024;    // 10 MB
    private static final int CONNECT_TIMEOUT = 15;
    private static final int WRITE_TIMEOUT = 60;
    private static final int TIMEOUT = 60;

    public static OkHttpClient provideOkHttp(Application application) {
        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        Cache cache = new Cache(application.getCacheDir(), CACHE_SIZE);

        return new OkHttpClient.Builder()
                .connectTimeout(CONNECT_TIMEOUT, TimeUnit.SECONDS)
                .writeTimeout(WRITE_TIMEOUT, TimeUnit.SECONDS)
                .readTimeout(TIMEOUT, TimeUnit.SECONDS)
                .addInterceptor(loggingInterceptor)
                .cache(cache)
                .build();
    }

    public static Retrofit provideRetrofit(OkHttpClient okHttpClient) {
        return new Retrofit.Builder()
                .baseUrl(Values.URL_BASE)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createAsync())
                .client(okHttpClient)
                .build();
    }

    public static MoviesAPI provideAPI(Application application) {
        return provideRetrofit(provideOkHttp(application)).create(MoviesAPI.class);
    }
}

File: app/src/main/java/com/ngengs/android/popularmovies/apps/utils/networks/NetworkHelpers.kt
package com.ngengs.android.popularmovies.apps.utils.networks

import android.content.Context
import com.ngengs.android.popularmovies.apps.BuildConfig
import com.ngengs.android.popularmovies.apps.globals.Values
import okhttp3.Cache
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit


/**
 * Created by rizky.kharisma on 09/11/22.
 * @ngengs
 */
object NetworkHelpers {
    private const val CACHE_SIZE = 10 * 1024 * 1024L  // 10 MB
    private const val CONNECT_TIMEOUT = 15
    private const val WRITE_TIMEOUT = 60
    private const val TIMEOUT = 60

    private fun provideOkHttp(context: Context): OkHttpClient {
        val loggingInterceptor = HttpLoggingInterceptor()
        if (BuildConfig.DEBUG) loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY)

        val cache = Cache(context.cacheDir, CACHE_SIZE)

        return OkHttpClient.Builder()
            .connectTimeout(CONNECT_TIMEOUT.toLong(), TimeUnit.SECONDS)
            .writeTimeout(WRITE_TIMEOUT.toLong(), TimeUnit.SECONDS)
            .readTimeout(TIMEOUT.toLong(), TimeUnit.SECONDS)
            .addInterceptor(loggingInterceptor)
            .cache(cache)
            .build()
    }

    private fun provideRetrofit(okHttpClient: OkHttpClient): Retrofit = Retrofit.Builder()
        .baseUrl(Values.URL_BASE)
        .addConverterFactory(GsonConverterFactory.create(GsonConfig.gson))
        .addCallAdapterFactory(RxJava2CallAdapterFactory.createAsync())
        .client(okHttpClient)
        .build()

    fun provideAPI(context: Context): MoviesAPI {
        return provideRetrofit(provideOkHttp(context)).create(MoviesAPI::class.java)
    }
}